influxdb_iox/src/influxdb_ioxd/server_type/test.rs

```diff
@@ -0,0 +1,127 @@
+use hyper::{Body, Method, Request, Response};
+use metric::Registry;
+use snafu::Snafu;
+use tokio_util::sync::CancellationToken;
+use trace::TraceCollector;
+
+use super::{RpcError, ServerType};
+
+#[derive(Debug, Snafu)]
+pub enum ApplicationError {
+    #[snafu(display("No handler for {:?} {}", method, path))]
+    RouteNotFound { method: Method, path: String },
+}
+
+impl HttpApiErrorSource for ApplicationError {
+    fn to_http_api_error(&self) -> HttpApiError {
+        match self {
+            e @ Self::RouteNotFound { .. } => e.not_found(),
+        }
+    }
+}
+
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum TestAction {
+    None,
+    EarlyReturnFromGrpcWorker,
+    EarlyReturnFromServerWorker,
+    PanicInGrpcWorker,
+    PanicInServerWorker,
+}
+
+impl FromStr for TestAction {
+    type Err = String;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s.to_lowercase().as_ref() {
+            "none" => Ok(Self::None),
+            "early-return-from-grpc-worker" => Ok(Self::EarlyReturnFromGrpcWorker),
+            "early-return-from-server-worker" => Ok(Self::EarlyReturnFromServerWorker),
+            "panic-in-grpc-worker" => Ok(Self::PanicInGrpcWorker),
+            "panic-in-server-worker" => Ok(Self::PanicInServerWorker),
+            _ => Err(format!("Invalid test action: {}", s)),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub struct TestServerType {
+    metric_registry: Arc<Registry>,
+    trace_collector: Option<Arc<dyn TraceCollector>>,
+    shutdown: CancellationToken,
+    test_action: TestAction,
+}
+
+impl TestServerType {
+    pub fn new(
+        metric_registry: Arc<Registry>,
+        trace_collector: Option<Arc<dyn TraceCollector>>,
+        test_action: TestAction,
+    ) -> Self {
+        Self {
+            metric_registry,
+            trace_collector,
+            shutdown: CancellationToken::new(),
+            test_action,
+        }
+    }
+}
+
+#[async_trait]
+impl ServerType for TestServerType {
+    type RouteError = ApplicationError;
+
+    fn metric_registry(&self) -> Arc<Registry> {
+        Arc::clone(&self.metric_registry)
+    }
+
+    fn trace_collector(&self) -> Option<Arc<dyn TraceCollector>> {
+        self.trace_collector.clone()
+    }
+
+    async fn route_http_request(
+        &self,
+        req: Request<Body>,
+    ) -> Result<Response<Body>, Self::RouteError> {
+        Err(ApplicationError::RouteNotFound {
+            method: req.method().clone(),
+            path: req.uri().path().to_string(),
+        })
+    }
+
+    async fn server_grpc(self: Arc<Self>, builder_input: RpcBuilderInput) -> Result<(), RpcError> {
+        if self.test_action == TestAction::PanicInGrpcWorker {
+            panic!("Test panic in gRPC worker");
+        }
+        if self.test_action == TestAction::EarlyReturnFromGrpcWorker {
+            return Ok(());
+        }
+
+        let builder = setup_builder!(builder_input, self);
+        serve_builder!(builder);
+
+        Ok(())
```

Comment:
Personally, I think these functions with multiple `if` blocks would be better expressed with `match`:

```suggestion
        match self.test_action {
            TestAction::PanicInGrpcWorker => panic!("Test panic in gRPC worker"),
            TestAction::EarlyReturnFromGrpcWorker => Ok(()),
            _ => {
                let builder = setup_builder!(builder_input, self);
                serve_builder!(builder);

                Ok(())
            }
        }
```

but I don't feel super strongly about it, especially because this is test code :)